eclipse-collections/src/main/java/org/eclipse/collections/impl/multimap/AbstractSynchronizedMultimap.java

```diff
@@ -0,0 +1,399 @@
+    {
+        synchronized (this.lock)
+        {
+            return this.getDelegate().putAll(key, values);
+        }
+    }
+
+    @Override
+    public <KK extends K, VV extends V> boolean putAll(Multimap<KK, VV> multimap)
+    {
+        synchronized (this.lock)
+        {
+            return this.getDelegate().putAll(multimap);
+        }
+    }
+
+    @Override
+    public void clear()
+    {
+        synchronized (this.lock)
+        {
+            this.getDelegate().clear();
+        }
+    }
+
+    @Override
+    public boolean isEmpty()
+    {
+        synchronized (this.lock)
+        {
+            return this.delegate.isEmpty();
+        }
+    }
+
+    @Override
+    public boolean notEmpty()
+    {
+        synchronized (this.lock)
+        {
+            return this.delegate.notEmpty();
+        }
+    }
+
+    @Override
+    public void forEachValue(Procedure<? super V> procedure)
+    {
+        synchronized (this.lock)
+        {
+            this.delegate.forEachValue(procedure);
+        }
+    }
+
+    @Override
+    public void forEachKey(Procedure<? super K> procedure)
+    {
+        synchronized (this.lock)
+        {
+            this.delegate.forEachKey(procedure);
+        }
+    }
+
+    @Override
+    public void forEachKeyValue(Procedure2<? super K, ? super V> procedure)
+    {
+        synchronized (this.lock)
+        {
+            this.delegate.forEachKeyValue(procedure);
+        }
+    }
+
+    @Override
+    public void forEachKeyMultiValues(Procedure2<? super K, ? super Iterable<V>> procedure)
+    {
+        synchronized (this.lock)
+        {
+            this.delegate.forEachKeyMultiValues(procedure);
+        }
+    }
+
+    @Override
+    public int size()
+    {
+        synchronized (this.lock)
+        {
+            return this.delegate.size();
+        }
+    }
+
+    @Override
+    public int sizeDistinct()
+    {
+        synchronized (this.lock)
+        {
+            return this.delegate.sizeDistinct();
+        }
+    }
+
+    @Override
+    public boolean containsKey(Object key)
+    {
+        synchronized (this.lock)
+        {
+            return this.delegate.containsKey(key);
+        }
+    }
+
+    @Override
+    public boolean containsValue(Object value)
+    {
+        synchronized (this.lock)
+        {
+            return this.delegate.containsValue(value);
+        }
+    }
+
+    @Override
+    public boolean containsKeyAndValue(Object key, Object value)
+    {
+        synchronized (this.lock)
+        {
+            return this.delegate.containsKeyAndValue(key, value);
+        }
+    }
+
+    @Override
+    public RichIterable<K> keysView()
+    {
+        synchronized (this.lock)
+        {
+            return this.delegate.keysView();
+        }
+    }
+
+    @Override
+    public SetIterable<K> keySet()
+    {
+        synchronized (this.lock)
+        {
+            return this.delegate.keySet();
+        }
```

Comment:
keySet() must not take linear time. Why not just return the delegate keySet
wrapped in an unmodifiable wrapper and rewrapped in a synchronized wrapper
that shares a lock with the multimap?
On Sun, Aug 14, 2016 at 6:46 AM Hiroshi Ito <notifications@github.com>
wrote:

> In
> eclipse-collections/src/main/java/org/eclipse/collections/impl/multimap/AbstractSynchronizedMultimap.java
> <https://github.com/eclipse/eclipse-collections/pull/93#discussion_r74699709>
> :
>
> > +    @Override
> > +    public RichIterable<K> keysView()
> > +    {
> > +        synchronized (this.lock)
> > +        {
> > +            return this.delegate.keysView();
> > +        }
> > +    }
> > +
> > +    @Override
> > +    public SetIterable<K> keySet()
> > +    {
> > +        synchronized (this.lock)
> > +        {
> > +            return this.delegate.keySet();
> > +        }
>
> Thanks for making a good progress on this.
>
> As for (1), I suggest we modify AbstractMultimapTestCase#keySet to check
> it throw UnsupportedOperationException when being mutated, instead of
> asserting the keySet being instance of Unmodifiable. I think you can just
> make the delegate's keySet unmodifiable to make it pass.
>
> (2) looks to be a bit too much to address in this change... I suggest we
> compromise O(1) requirement for synchronized multimaps keySet and make them
> O(N) for now (with documenting that in Javadoc). I'd suggest to raise
> another issue to improve this performance, that requires a design
> discussion and potentially a good amount of refactoring.
>
> @motlin <https://github.com/motlin> , @nikhilnanivadekar
> <https://github.com/nikhilnanivadekar> , what do you think?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/eclipse/eclipse-collections/pull/93/files/ace474519c1dbe03d2783ec5640c911f58ada19d..7de3f8ba027df68f6a4eb3621877b9007dbe0c11#r74699709>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAO6ImGCQLLyfnQc1ipJtWa8InGp_LjEks5qfvIUgaJpZM4ILYAA>
> .
>